packages/parse5/lib/tokenizer/index.ts

```diff
@@ -220,17 +220,22 @@ export class Tokenizer {
 
     private consumedAfterSnapshot = -1;
 
+    private ctLoc: Location | null;
     private currentCharacterToken: CharacterToken | null = null;
     private currentToken: Token | null = null;
     private currentAttr: Attribute = { name: '', value: '' };
 
+    // NOTE: Doctypes tokens are created with several different offset. We keep track of the moment a doctype *might* start here.
```

Comment:
Weird? Should examples of differences perhaps be documented here?

Reply:
It just occurred to me that the `ctLoc` property can be used here; updated accordingly.

The issue with doctypes is that they don't start at fixed offsets. Eg. whitespace can push the token creation further back.

Reply:
how is that different from elements, or anything else, that is also pushed back by whitespace?

Reply:
Doctype tokens are a bit weird in the way they are created in the spec. All other tokens are created as soon as you can be certain that a specific type of token is needed. With Doctypes, this is pushed back until we get the beginning of something that needs to be saved. Eg. `<!doctype         html>` would only have its token created on the `h` of `html`.

Reply:
Do you have a link to the spec that mentions this behavior?
I don’t understand, from the tokenizing chapter, what makes `DOCTYPE`, `[CDATA[`, or `--` different?

Reply:
The [Before DOCTYPE name state](https://html.spec.whatwg.org/multipage/parsing.html#before-doctype-name-state) is probably the best example here. The token is only created if a non-whitespace character is found.

This is in contrast to the [Markup declaration open state](https://html.spec.whatwg.org/multipage/parsing.html#markup-declaration-open-state), which creates comment tokens after reading a specific number of characters. This allows us to set specific offsets. The same applies to the [Tag open state](https://html.spec.whatwg.org/multipage/parsing.html#tag-open-state) and all of the `* end tag open state`s. These are all the states in which non-doctype tokens are created.